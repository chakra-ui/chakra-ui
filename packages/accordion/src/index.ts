export * from "./accordion"
export * from "./use-accordion"
