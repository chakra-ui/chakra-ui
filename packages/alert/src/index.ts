export { Alert, type AlertProps } from "./alert"
export { useAlertStyles, type AlertStatus } from "./alert-context"
export {
  AlertDescription,
  type AlertDescriptionProps,
} from "./alert-description"
export { AlertIcon, type AlertIconProps } from "./alert-icon"
export { AlertTitle, type AlertTitleProps } from "./alert-title"
