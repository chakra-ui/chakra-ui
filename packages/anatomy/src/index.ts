import { anatomy } from "@chakra-ui/theme-tools"

/**
 * **Accordion anatomy**
 * - Root: the root container of the accordion
 * - Container: the accordion item contains the button and panel
 * - Button: the button is the trigger for the panel
 * - Panel: the panel is the content of the accordion item
 * - Icon: the expanded/collapsed icon
 */
export const accordionAnatomy = anatomy("accordion")
  .parts("root", "container", "button", "panel")
  .extend("icon")

/**
 * **Alert anatomy**
 * - Title: the alert's title
 * - Description: the alert's description
 * - Icon: the alert's icon
 */
export const alertAnatomy = anatomy("alert")
  .parts("title", "description", "container")
  .extend("icon")

/**
 * **Avatar anatomy**
 * - Container: the container for the avatar
 * - Label: the avatar initials text
 * - Excess Label: the label or text that represents excess avatar count.
 * Typically used in avatar groups.
 * - Group: the container for the avatar group
 */
export const avatarAnatomy = anatomy("avatar")
  .parts("label", "badge", "container")
  .extend("excessLabel", "group")

/**
 * **Breadcrumb anatomy**
 * - Item: the container for a breadcrumb item
 * - Link: the element that represents the breadcrumb link
 * - Container: the container for the breadcrumb items
 * - Separator: the separator between breadcrumb items
 */
export const breadcrumbAnatomy = anatomy("breadcrumb")
  .parts("link", "item", "container")
  .extend("separator")

export const buttonAnatomy = anatomy("button").parts()

export const checkboxAnatomy = anatomy("checkbox")
  .parts("control", "icon", "container")
  .extend("label")

export const circularProgressAnatomy = anatomy("progress")
  .parts("track", "filledTrack")
  .extend("label")

export const drawerAnatomy = anatomy("drawer")
  .parts("overlay", "dialogContainer", "dialog")
  .extend("header", "closeButton", "body", "footer")

export const editableAnatomy = anatomy("editable").parts(
  "preview",
  "input",
  "textarea",
)

export const formAnatomy = anatomy("form").parts(
  "container",
  "requiredIndicator",
  "helperText",
)

export const formErrorAnatomy = anatomy("formError").parts("text", "icon")

export const inputAnatomy = anatomy("input").parts("addon", "field", "element")

export const listAnatomy = anatomy("list").parts("container", "item", "icon")

export const menuAnatomy = anatomy("menu")
  .parts("button", "list", "item")
  .extend("groupTitle", "command", "divider")

export const modalAnatomy = anatomy("modal")
  .parts("overlay", "dialogContainer", "dialog")
  .extend("header", "closeButton", "body", "footer")

export const numberInputAnatomy = anatomy("numberinput").parts(
  "root",
  "field",
  "stepperGroup",
  "stepper",
)

export const pinInputAnatomy = anatomy("pininput").parts("field")

export const popoverAnatomy = anatomy("popover")
  .parts("content", "header", "body", "footer")
  .extend("popper", "arrow", "closeButton")

export const progressAnatomy = anatomy("progress").parts(
  "label",
  "filledTrack",
  "track",
)

export const radioAnatomy = anatomy("radio").parts(
  "container",
  "control",
  "label",
)

export const selectAnatomy = anatomy("select").parts("field", "icon")

export const sliderAnatomy = anatomy("slider").parts(
  "container",
  "track",
  "thumb",
  "filledTrack",
)

export const statAnatomy = anatomy("stat").parts(
  "container",
  "label",
  "helpText",
  "number",
  "icon",
)

export const switchAnatomy = anatomy("switch").parts(
  "container",
  "track",
  "thumb",
)

export const tableAnatomy = anatomy("table").parts(
  "table",
  "thead",
  "tbody",
  "tr",
  "th",
  "td",
  "tfoot",
  "caption",
)

export const tabsAnatomy = anatomy("tabs").parts(
  "root",
  "tab",
  "tablist",
  "tabpanel",
  "tabpanels",
  "indicator",
)

/**
 * **Tag anatomy**
 * - Container: the container for the tag
 * - Label: the text content of the tag
 * - closeButton: the close button for the tag
 */
export const tagAnatomy = anatomy("tag").parts(
  "container",
  "label",
  "closeButton",
)
