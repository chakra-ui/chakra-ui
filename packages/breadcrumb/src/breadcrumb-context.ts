import { createContext } from "@chakra-ui/react-utils"
import { SystemStyleObject } from "@chakra-ui/system"
import { Dict } from "@chakra-ui/utils"

export const [BreadcrumbStylesProvider, useBreadcrumbStyles] = createContext<
  Dict<SystemStyleObject>
>({
  name: `BreadcrumbStylesContext`,
  errorMessage: `useBreadcrumbStyles returned is 'undefined'. Seems you forgot to wrap the components in "<Breadcrumb />" `,
})
