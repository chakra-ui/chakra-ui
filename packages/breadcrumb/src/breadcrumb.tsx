import {
  chakra,
  forwardRef,
  PropsOf,
  useMultiStyleConfig,
  StylesProvider,
  useStyles,
  SystemProps,
  ThemingProps,
  omitThemingProps,
} from "@chakra-ui/system"
import { cx, getValidChildren, __DEV__ } from "@chakra-ui/utils"
import React, { Ref, cloneElement } from "react"

export type BreadcrumbSeparatorProps = PropsOf<typeof chakra.div> & {
  spacing?: SystemProps["mx"]
}

/**
 * React component that separates each breadcrumb link
 */
export const BreadcrumbSeparator = React.forwardRef(
  function BreadcrumbSeparator(props: BreadcrumbSeparatorProps, ref: Ref<any>) {
    const { spacing, ...rest } = props

    const styles = useStyles()
    const separatorStyles = {
      mx: spacing,
      ...styles.separator,
    }

    return (
      <chakra.span
        ref={ref}
        role="presentation"
        {...rest}
        __css={separatorStyles}
      />
    )
  },
)

if (__DEV__) {
  BreadcrumbSeparator.displayName = "BreadcrumbSeparator"
}

interface LinkOptions {
  isCurrentPage?: boolean
}

export type BreadcrumbLinkProps = PropsOf<typeof chakra.a> & LinkOptions

/**
 * Breadcrumb link.
 *
 * It renders a `span` when it's the current link. Otherwise,
 * it renders an anchor tag.
 */
export const BreadcrumbLink = forwardRef<BreadcrumbLinkProps>(
  function BreadcrumbLink(props, ref) {
    const { isCurrentPage, as, className, ...rest } = props
    const styles = useStyles()

    const sharedProps = {
      ref,
      as,
      className: cx("chakra-breadcrumb__link", className),
      ...rest,
    }

    if (isCurrentPage) {
      return <chakra.span aria-current="page" {...sharedProps} />
    }

    return <chakra.a __css={styles.link} {...sharedProps} />
  },
)

if (__DEV__) {
  BreadcrumbLink.displayName = "BreadcrumbLink"
}

type BreadcrumbItemOptions = BreadcrumbProps & {
  isCurrentPage?: boolean
  isLastChild?: boolean
}

export type BreadcrumbItemProps = BreadcrumbItemOptions &
  PropsOf<typeof chakra.li>

/**
 * BreadcrumbItem is used to group a breadcrumb link.
 * It renders a `li` element to denote it belongs to an order list of links.
 * @see Docs https://chakra-ui.com/components/breadcrumbs
 */
export const BreadcrumbItem = forwardRef<BreadcrumbItemProps>(
  function BreadcrumbItem(props, ref) {
    const {
      isCurrentPage,
      separator,
      isLastChild,
      spacing,
      children,
      className,
      ...rest
    } = props

    const validChildren = getValidChildren(children)

    const clones = validChildren.map((child) => {
      if (child.type === BreadcrumbLink) {
        return cloneElement(child, {
          isCurrentPage,
        })
      }

      if (child.type === BreadcrumbSeparator) {
        return cloneElement(child, {
          spacing,
          children: child.props.children || separator,
        })
      }

      return child
    })

    const _className = cx("chakra-breadcrumb__list-item", className)

    return (
      <chakra.li
        ref={ref}
        className={_className}
        {...rest}
        __css={{
          display: "inline-flex",
          alignItems: "center",
        }}
      >
        {clones}
        {!isLastChild && (
          <BreadcrumbSeparator spacing={spacing} children={separator} />
        )}
      </chakra.li>
    )
  },
)

if (__DEV__) {
  BreadcrumbItem.displayName = "BreadcrumbItem"
}

export interface BreadcrumbOptions {
  children?: React.ReactNode
  /**
   * The visual separator between each breadcrumb item
   */
  separator?: string | React.ReactElement
  /**
   * The left and right margin applied to the separator
   */
  spacing?: SystemProps["mx"]
}

export type BreadcrumbProps = PropsOf<typeof chakra.nav> &
  BreadcrumbOptions &
  ThemingProps

/**
 * Breadcrumb is used to render a breadcrumb navigation landmark.
 * It renders a `nav` element with `aria-label` set to `Breadcrumb`
 *
 * @see Docs https://chakra-ui.com/components/breadcrumbs
 */
export const Breadcrumb = React.forwardRef(function Breadcrumb(
  props: BreadcrumbProps,
  ref: Ref<any>,
) {
  const styles = useMultiStyleConfig("Breadcrumb", props)
  const realProps = omitThemingProps(props)

  const {
    children,
    spacing = "0.5rem",
    separator = "/",
    className,
    ...rest
  } = realProps

  const validChildren = getValidChildren(children)
  const count = validChildren.length

  const clones = validChildren.map((child, index) =>
    cloneElement(child, {
      separator,
      spacing,
      isLastChild: count === index + 1,
    }),
  )

  const _className = cx("chakra-breadcrumb", className)

  return (
    <chakra.nav
      ref={ref}
      aria-label="breadcrumb"
      className={_className}
      {...rest}
    >
      <StylesProvider value={styles}>
        <chakra.ol className="chakra-breadcrumb__list">{clones}</chakra.ol>
      </StylesProvider>
    </chakra.nav>
  )
})

if (__DEV__) {
  Breadcrumb.displayName = "Breadcrumb"
}
