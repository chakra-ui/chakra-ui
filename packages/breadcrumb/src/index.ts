export { Breadcrumb, type BreadcrumbProps } from "./breadcrumb"
export { BreadcrumbItem, type BreadcrumbItemProps } from "./breadcrumb-item"
export {
  BreadcrumbSeparator,
  type BreadcrumbSeparatorProps,
} from "./breadcrumb-separator"
export { BreadcrumbLink, type BreadcrumbLinkProps } from "./breadcrumb-link"
export { useBreadcrumbStyles } from "./breadcrumb-context"
export type { BreadcrumbOptions } from "./breadcrumb-types"
