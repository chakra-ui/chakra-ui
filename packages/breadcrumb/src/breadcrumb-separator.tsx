import {
  chakra,
  forwardRef,
  HTMLChakraProps,
  SystemStyleObject,
} from "@chakra-ui/system"
import { useBreadcrumbStyles } from "./breadcrumb-context"

export interface BreadcrumbSeparatorProps extends HTMLChakraProps<"div"> {
  /**
   * @type SystemStyleObject["mx"]
   */
  spacing?: SystemStyleObject["mx"]
}
/**
 * React component that separates each breadcrumb link
 */

export const BreadcrumbSeparator = forwardRef<BreadcrumbSeparatorProps, "span">(
  function BreadcrumbSeparator(props, ref) {
    const { spacing, ...rest } = props

    const styles = useBreadcrumbStyles()
    const separatorStyles: SystemStyleObject = {
      mx: spacing,
      ...styles.separator,
    }

    return (
      <chakra.span
        ref={ref}
        role="presentation"
        {...rest}
        __css={separatorStyles}
      />
    )
  },
)

BreadcrumbSeparator.displayName = "BreadcrumbSeparator"
