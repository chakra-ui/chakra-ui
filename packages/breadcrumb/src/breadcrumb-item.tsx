import {
  chakra,
  forwardRef,
  SystemStyleObject,
  HTMLChakraProps,
} from "@chakra-ui/system"
import { getValidChildren } from "@chakra-ui/react-children-utils"
import { cx } from "@chakra-ui/shared-utils"
import { useBreadcrumbStyles } from "./breadcrumb-context"
import { BreadcrumbSeparator } from "./breadcrumb-separator"
import { BreadcrumbLink } from "./breadcrumb-link"
import { cloneElement } from "react"
import { BreadcrumbItemOptions } from "./breadcrumb-types"

export interface BreadcrumbItemProps
  extends BreadcrumbItemOptions,
    HTMLChakraProps<"li"> {}

/**
 * BreadcrumbItem is used to group a breadcrumb link.
 * It renders a `li` element to denote it belongs to an order list of links.
 *
 * @see Docs https://chakra-ui.com/breadcrumb
 */

export const BreadcrumbItem = forwardRef<BreadcrumbItemProps, "li">(
  function BreadcrumbItem(props, ref) {
    const {
      isCurrentPage,
      separator,
      isLastChild,
      spacing,
      children,
      className,
      ...rest
    } = props

    const validChildren = getValidChildren(children)

    const clones = validChildren.map((child) => {
      if (child.type === BreadcrumbLink) {
        return cloneElement(child, {
          isCurrentPage,
        })
      }

      if (child.type === BreadcrumbSeparator) {
        return cloneElement(child, {
          spacing,
          children: child.props.children || separator,
        })
      }

      return child
    })

    const styles = useBreadcrumbStyles()
    const itemStyles: SystemStyleObject = {
      display: "inline-flex",
      alignItems: "center",
      ...styles.item,
    }

    const _className = cx("chakra-breadcrumb__list-item", className)

    return (
      <chakra.li ref={ref} className={_className} {...rest} __css={itemStyles}>
        {clones}
        {!isLastChild && (
          <BreadcrumbSeparator spacing={spacing}>
            {separator}
          </BreadcrumbSeparator>
        )}
      </chakra.li>
    )
  },
)
BreadcrumbItem.displayName = "BreadcrumbItem"
