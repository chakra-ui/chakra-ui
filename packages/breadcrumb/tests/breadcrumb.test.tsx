import * as React from "react"
import { render, screen, testA11y } from "@chakra-ui/test-utils"
import { Breadcrumb, BreadcrumbItem, BreadcrumbLink } from "../src"

it("passes a11y test", async () => {
  await testA11y(
    <Breadcrumb>
      <BreadcrumbItem>
        <BreadcrumbLink href="#">Link 1</BreadcrumbLink>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <BreadcrumbLink href="#">Link 2</BreadcrumbLink>
      </BreadcrumbItem>
      <BreadcrumbItem isCurrentPage>
        <BreadcrumbLink>Link 3</BreadcrumbLink>
      </BreadcrumbItem>
    </Breadcrumb>,
  )
})

test("has the proper aria-attributes", () => {
  render(
    <Breadcrumb>
      <BreadcrumbItem>
        <BreadcrumbLink href="#">Link 1</BreadcrumbLink>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <BreadcrumbLink href="#">Link 2</BreadcrumbLink>
      </BreadcrumbItem>
      <BreadcrumbItem isCurrentPage>
        <BreadcrumbLink>Link 3</BreadcrumbLink>
      </BreadcrumbItem>
    </Breadcrumb>,
  )

  // surrounding `nav` has aria-label="breadcrumb"
  screen.getByLabelText("breadcrumb", { selector: "nav" })

  // `isCurrentPage` link has aria-current="page"
  const currentPageLink = screen.getByText("Link 3")
  expect(currentPageLink).toHaveAttribute("aria-current", "page")

  // separator receives presentation="role"
  expect(screen.getAllByRole("presentation")).toHaveLength(2)
})

test("separator can be changed", () => {
  render(
    <Breadcrumb separator="-">
      <BreadcrumbItem>
        <BreadcrumbLink href="#">Link 1</BreadcrumbLink>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <BreadcrumbLink href="#">Link 2</BreadcrumbLink>
      </BreadcrumbItem>
    </Breadcrumb>,
  )
  expect(screen.getAllByText("-")).toHaveLength(1)
})
