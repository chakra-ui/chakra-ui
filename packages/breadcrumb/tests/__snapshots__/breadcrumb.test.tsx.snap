// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb renders correctly 1`] = `
<DocumentFragment>
  .emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  -webkit-transition: all 0.15s ease-out;
  transition: all 0.15s ease-out;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
  color: inherit;
}

.emotion-0:hover,
.emotion-0[data-hover] {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:focus,
.emotion-0[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-1 {
  margin-left: 0.5rem;
  margin-right: 0.5rem;
}

<nav
    aria-label="breadcrumb"
    class="chakra-breadcrumb emotion-6"
  >
    <ol
      class="chakra-breadcrumb__list emotion-6"
    >
      <li
        class="chakra-breadcrumb__list-item emotion-2"
      >
        <a
          class="chakra-breadcrumb__link emotion-0"
          href="#"
        >
          Link 1
        </a>
        <span
          class="emotion-1"
          role="presentation"
        >
          /
        </span>
      </li>
      <li
        class="chakra-breadcrumb__list-item emotion-2"
      >
        <a
          class="chakra-breadcrumb__link emotion-0"
          href="#"
        >
          Link 2
        </a>
        <span
          class="emotion-1"
          role="presentation"
        >
          /
        </span>
      </li>
      <li
        class="chakra-breadcrumb__list-item emotion-2"
      >
        <span
          aria-current="page"
          class="chakra-breadcrumb__link emotion-6"
        >
          Link 3
        </span>
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;
